File: frontend/src/mocks/pong/simulation/collisions.ts
import type { Side } from "../../../games/pong/protocol/types";
import { clamp } from "../../../games/pong/shared/math";
import {
  ANGLE_FACTOR,
  BASE_PADDLE_SPEED,
  BORDERS,
  PADDLE_MOTION_FACTOR,
  SIDES,
  TABLE,
  type Border,
} from "./constants";
import {
  borderSign,
  getPaddleOffset,
  sideSign,
} from "./geometry";
import { getPaddleVelocity } from "./paddles";
import type { SimState } from "./types";

/**
 * Verifica se a bola e a paddle se sobrepõem no eixo Z (largura).
 *
 * @param paddleOffset - Posição da paddle no eixo Z (largura)
 * @param ballZ - Posição da bola no eixo Z (largura)
 */
function doesPaddleOverlapBall(paddleOffset: number, ballZ: number): boolean {
  const overlapDistance = TABLE.paddleHalfDepth + TABLE.ballRadius;
  return Math.abs(paddleOffset - ballZ) <= overlapDistance;
}

/**
 * Verifica se a velocidade no eixo X (comprimento) aponta para o lado informado.
 */
function isMovingTowardSide(velocityX: number, side: Side): boolean {
  return side === "left" ? velocityX < 0 : velocityX > 0;
}

/**
 * Verifica se a bola está na faixa de colisão da paddle no eixo X (comprimento).
 */
function isBallInPaddleHitZone(ballX: number, side: Side): boolean {
  const paddleFaceX = sideSign(side) * TABLE.paddleX;
  const minX = paddleFaceX - TABLE.ballRadius;
  const maxX = paddleFaceX + TABLE.ballRadius;
  return ballX >= minX && ballX <= maxX;
}

/**
 * Verifica se a bola está colidindo com a paddle do lado informado.
 */
function isHittingPaddle(state: SimState, side: Side): boolean {
  const movingTowardPaddle = isMovingTowardSide(state.velocityX, side);
  const insideHitZone = isBallInPaddleHitZone(state.ballX, side);
  const overlapsPaddle = doesPaddleOverlapBall(
    getPaddleOffset(state, side),
    state.ballZ,
  );
  return movingTowardPaddle && insideHitZone && overlapsPaddle;
}

/**
 * Verifica se a bola ultrapassou a borda superior ou inferior no eixo Z (largura).
 */
function isBallPastBorder(ballZ: number, border: Border): boolean {
  const limit = borderSign(border) * (TABLE.halfWidth - TABLE.ballRadius);
  return border === "top" ? ballZ > limit : ballZ < limit;
}

/**
 * Reflete a bola na borda e corrige a posição no eixo Z (largura).
 */
function bounceOnBorder(state: SimState, border: Border): void {
  state.ballZ = borderSign(border) * (TABLE.halfWidth - TABLE.ballRadius);
  state.velocityZ *= -1;
}

/**
 * Calcula o ponto de contato na paddle no eixo Z (largura), normalizado entre -1 e 1.
 * Valores próximos de ±1 indicam contato perto das extremidades da paddle.
 */
function getPaddleHitRatio(state: SimState, side: Side): number {
  const paddleOffset = getPaddleOffset(state, side);
  const rawRatio = (state.ballZ - paddleOffset) / TABLE.paddleHalfDepth;
  return clamp(rawRatio, -1, 1);
}

/**
 * Posiciona a bola na face da paddle após a colisão, no eixo X (comprimento).
 */
function placeBallOnPaddleFace(state: SimState, side: Side): void {
  const paddleFaceX = sideSign(side) * TABLE.paddleX;
  state.ballX = paddleFaceX - sideSign(side) * TABLE.ballRadius;
}

/**
 * Define a velocidade da bola após o impacto com a paddle.
 * A componente Z depende do ponto de contato e da velocidade da paddle.
 *
 * @param hitRatio - Ponto de contato normalizado no eixo Z (largura) (−1 a 1)
 */
function applyPaddleBounceVelocity(
  state: SimState,
  side: Side,
  hitRatio: number,
): void {
  const directionX = side === "left" ? 1 : -1;
  const paddleMotion = getPaddleVelocity(state, side) * PADDLE_MOTION_FACTOR;

  state.velocityX = directionX * BASE_PADDLE_SPEED;
  state.velocityZ = hitRatio * ANGLE_FACTOR + paddleMotion;
}

/** Aplica o rebote completo da bola na paddle. */
function bounceOnPaddle(state: SimState, side: Side): void {
  const hitRatio = getPaddleHitRatio(state, side);
  placeBallOnPaddleFace(state, side);
  applyPaddleBounceVelocity(state, side, hitRatio);
}

/** Trata colisões da bola com as bordas superior e inferior. */
export function resolveSideBorders(state: SimState): void {
  for (const border of BORDERS) {
    if (!isBallPastBorder(state.ballZ, border)) continue;
    bounceOnBorder(state, border);
    return;
  }
}

/** Trata colisões da bola com as paddles. */
export function resolvePaddleCollisions(state: SimState): void {
  for (const side of SIDES) {
    if (!isHittingPaddle(state, side)) continue;
    bounceOnPaddle(state, side);
    return;
  }
}

File: backend/app/games/pong/collisions.py

from .constants import (
    ANGLE_FACTOR,
    BASE_PADDLE_SPEED,
    BORDERS,
    PADDLE_MOTION_FACTOR,
    SIDES,
    TABLE,
    border_sign,
    side_sign,
)
from .paddles import get_paddle_offset, get_paddle_velocity
from .shared import clamp
from .state import SimState

def does_paddle_overlap_ball(paddle_offset: float, ball_z: float) -> bool:
    overlap_distance = TABLE["paddleHalfDepth"] + TABLE["ballRadius"]
    return abs(paddle_offset - ball_z) <= overlap_distance

def is_moving_toward_side(velocity_x: float, side: str) -> bool:
    return velocity_x < 0 if side == "left" else velocity_x > 0

def is_ball_in_paddle_hit_zone(ball_x: float, side: str) -> bool:
    paddle_face_x = side_sign(side) * TABLE["paddleX"]
    min_x = paddle_face_x - TABLE["ballRadius"]
    max_x = paddle_face_x + TABLE["ballRadius"]
    return min_x <= ball_x <= max_x

def is_hitting_paddle(state: SimState, side: str) -> bool:
    moving_toward_paddle = is_moving_toward_side(state.velocity_x, side)
    inside_hit_zone = is_ball_in_paddle_hit_zone(state.ball_x, side)
    overlaps_paddle = does_paddle_overlap_ball(
        get_paddle_offset(state, side), state.ball_z
    )
    return moving_toward_paddle and inside_hit_zone and overlaps_paddle

def is_ball_past_border(ball_z: float, border: str) -> bool:
    limit = border_sign(border) * (TABLE["halfWidth"] - TABLE["ballRadius"])
    return ball_z > limit if border == "top" else ball_z < limit

def bounce_on_border(state: SimState, border: str) -> None:
    state.ball_z = border_sign(border) * (TABLE["halfWidth"] - TABLE["ballRadius"])
    state.velocity_z *= -1

def get_paddle_hit_ratio(state: SimState, side: str) -> float:
    paddle_offset = get_paddle_offset(state, side)
    raw_ratio = (state.ball_z - paddle_offset) / TABLE["paddleHalfDepth"]
    return clamp(raw_ratio, -1, 1)

def place_ball_on_paddle_face(state: SimState, side: str) -> None:
    paddle_face_x = side_sign(side) * TABLE["paddleX"]
    state.ball_x = paddle_face_x - side_sign(side) * TABLE["ballRadius"]

def apply_paddle_bounce_velocity(state: SimState, side: str, hit_ratio: float) -> None:
    direction_x = 1 if side == "left" else -1
    paddle_motion = get_paddle_velocity(state, side) * PADDLE_MOTION_FACTOR

    state.velocity_x = direction_x * BASE_PADDLE_SPEED
    state.velocity_z = hit_ratio * ANGLE_FACTOR + paddle_motion

def bounce_on_paddle(state: SimState, side: str) -> None:
    hit_ratio = get_paddle_hit_ratio(state, side)
    place_ball_on_paddle_face(state, side)
    apply_paddle_bounce_velocity(state, side, hit_ratio)

def resolve_side_borders(state: SimState) -> None:
    for border in BORDERS:
        if not is_ball_past_border(state.ball_z, border):
            continue
        bounce_on_border(state, border)
        return

def resolve_paddle_collisions(state: SimState) -> None:
    for side in SIDES:
        if not is_hitting_paddle(state, side):
            continue
        bounce_on_paddle(state, side)
        return
